kafka-client/src/main/java/dev/responsive/kafka/internal/db/CassandraWindowedTable.java

```diff
@@ -428,47 +488,112 @@ public String name() {
   public WriterFactory<Stamped<Bytes>, SegmentPartition> init(
       final int kafkaPartition
   ) {
-    // TODO(sophie): store the stream-time in the offset metadata row and retrieve it to
-    //  initialize the state store's tracked stream-time and reserve the epoch for all
-    //  currently active segments
     final SegmentPartition metadataPartition = partitioner.metadataTablePartition(kafkaPartition);
 
-    client.execute(
+    final var initMetadata = client.execute(
         QueryBuilder.insertInto(name)
-            .value(PARTITION_KEY.column(), PARTITION_KEY.literal(metadataPartition.partitionKey))
+            .value(PARTITION_KEY.column(), PARTITION_KEY.literal(metadataPartition.tablePartition))
             .value(SEGMENT_ID.column(), SEGMENT_ID.literal(metadataPartition.segmentId))
             .value(ROW_TYPE.column(), METADATA_ROW.literal())
             .value(DATA_KEY.column(), DATA_KEY.literal(METADATA_KEY))
             .value(WINDOW_START.column(), WINDOW_START.literal(METADATA_TS))
             .value(OFFSET.column(), OFFSET.literal(NO_COMMITTED_OFFSET))
             .value(EPOCH.column(), EPOCH.literal(0L))
+            .value(STREAM_TIME.column(), STREAM_TIME.literal(UNINITIALIZED_STREAM_TIME))
             .ifNotExists()
             .build()
     );
 
-    return LwtWriterFactory.initialize(
+    if (initMetadata.wasApplied()) {
+      LOG.info("Created new metadata segment for kafka partition {}", kafkaPartition);
+    }
+
+    final long epoch = fetchEpoch(metadataPartition) + 1;
+    final var reserveMetadataEpoch = client.execute(reserveEpoch(metadataPartition, epoch));
+    if (!reserveMetadataEpoch.wasApplied()) {
+      handleEpochFencing(kafkaPartition, metadataPartition, epoch);
+    }
+
+    final long streamTime = fetchStreamTime(kafkaPartition);
+    kafkaPartitionToPendingFlushInfo.put(kafkaPartition, new PendingFlushInfo(streamTime));
+    LOG.info("Initialized stream-time to {} with epoch {} for kafka partition {}",
+             streamTime, epoch, kafkaPartition);
+
+    // since the active data segments depend on the current stream-time for the windowed table,
+    // which we won't know until we initialize it from the remote, the metadata like epoch and
+    // stream-time are stored in a special metadata partition/segment that's separate from the
+    // regular data partitions/segments and never expired
+    // therefore we initialize from the metadata partition and then broadcast the epoch to
+    // all the other partitions containing data for active segments
+    for (final SegmentPartition tablePartition : partitioner.activeSegments(kafkaPartition, streamTime)) {
+      final var reserveSegmentEpoch = client.execute(reserveEpoch(tablePartition, epoch));
+
+      if (!reserveSegmentEpoch.wasApplied()) {
+        handleEpochFencing(kafkaPartition, tablePartition, epoch);
+      } else {
+        LOG.info("SOPHIE: reserved epoch {} for kafka partition {} and segment {}",
```

Comment:
I think it could be helpful to have this log - perhaps as a summary at the end of all the epochs that it reserved? (certainly we should remove `SOPHIE:` though 😉 )

Reply:
hahahah whoops -- guess I missed removing all the logging from my debug sessions. Fixed